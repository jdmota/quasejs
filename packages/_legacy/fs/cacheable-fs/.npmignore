*.txt
__dev__
test
node_modules
yarn.lock
