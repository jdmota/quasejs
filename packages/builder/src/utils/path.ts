const path = require( "path" );
const slash = require( "slash" );

export const reExt = /\.(.+)$/;

export function lowerPath( p: string ) {
  return path.sep === "\\" ? p.toLowerCase() : p;
}

export function relative( id: string, cwd: string ): string {
  return slash( lowerPath( path.relative( cwd, id ) ) );
}

export function relativeURL( id: string, cwd: string ): string {
  return slash( lowerPath( path.relative( path.dirname( cwd ), id ) ) );
}

export function resolvePath( id: string, cwd: string ): string {
  return lowerPath( path.resolve( cwd, id ) );
}

export function getType( id: string ): string {
  const match = id.match( reExt );
  return match ? match[ 1 ] : "";
}

export function makeAbsolute( file: string ) {
  return lowerPath( path.resolve( process.cwd(), file ) );
}

export function isAbsolute( file: string ): boolean {
  return path.isAbsolute( file );
}
