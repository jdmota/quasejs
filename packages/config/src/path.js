// @flow
import { formatOption } from "./formating";

export default class Path {

  +arr: string[];
  where: ?string;

  constructor() {
    this.arr = [];
    this.where = null;
  }

  push( key: string | number ) {
    this.arr.push( key + "" );
  }

  pop() {
    this.arr.pop();
  }

  size() {
    return this.arr.length;
  }

  chainToString() {
    return this.arr.join( "." );
  }

  format() {
    return formatOption( `${this.arr.join( "." )}${this.where ? ` [in ${this.where}]` : ""}` );
  }

}
