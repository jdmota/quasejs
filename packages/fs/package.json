{
  "private": true,
  "name": "@quase/fs",
  "version": "0.1.0",
  "description": "FS",
  "license": "MIT",
  "homepage": "https://github.com/quase/quasejs",
  "repository": "https://github.com/quase/quasejs/tree/master/packages/fs",
  "main": "dist/index.js",
  "module": "src/index.js",
  "files": [
    "dist/index.js",
    "dist/index.js.map",
    "src"
  ],
  "dependencies": {
    "fs-extra": "^4.0.2",
    "lru-cache": "^4.1.1",
    "micromatch": "^2.3.11",
    "slash": "^1.0.0",
    "zen-observable": "^0.5.2"
  }
}
