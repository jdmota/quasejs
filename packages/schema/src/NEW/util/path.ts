export class Path {
  public readonly parent: Path | null;
  public readonly key: string | number | null;
  public readonly context: string | null;

  constructor(
    parent: Path,
    key: string | number | null,
    context: string | null
  );
  constructor(parent: null, key: null, context: null);
  constructor(
    parent: Path | null,
    key: string | number | null,
    context: string | null
  ) {
    this.parent = parent;
    this.key = key;
    this.context = context;
  }

  isRoot() {
    return this.parent == null;
  }

  push(key: string | number | null, context: string | null = null) {
    return new Path(this, key, context);
  }

  pop(): Path {
    if (this.parent == null)
      throw new Error("Invariant: validation path - cannot pop the root");
    return this.parent;
  }

  static create() {
    return new Path(null, null, null);
  }
}
