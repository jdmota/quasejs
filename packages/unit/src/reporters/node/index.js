import skipReasons from "../../core/skip-reasons";
import { log, log as printLog, logEol, indentString } from "./log";

const turbocolor = require( "turbocolor" );
const logSymbols = require( "log-symbols" );
const ora = require( "ora" );
const codeFrameColumns = require( "@babel/code-frame" ).codeFrameColumns;
const { prettify } = require( "@quase/path-url" );

export default class NodeReporter {

  constructor( runner ) {
    this.runner = runner;
    this.spinner = null;
    this.pendingTests = new Set();
    this.tests = [];
    this.otherErrors = [];
    this.ended = false;
    this.didAfterRun = false;
    this.interrupted = false;
    runner.on( "runStart", this.runStart.bind( this ) );
    runner.on( "testStart", this.testStart.bind( this ) );
    runner.on( "testEnd", this.testEnd.bind( this ) );
    runner.on( "runEnd", this.runEnd.bind( this ) );
    runner.on( "otherError", err => {
      this.otherErrors.push( err );
      if ( this.ended ) {
        this.logOtherErrors();
      }
    } );
    runner.on( "exit", async() => {
      NodeReporter.showSeed( runner );

      await this.logOtherErrors();

      if ( this.interrupted ) {
        log( `\n${turbocolor.bold.red( "Interrupted." )}\n\n` );
      }

      if ( process.exitCode ) {
        log( turbocolor.bold.red( "Exit code: " + process.exitCode ) );
      } else {
        log( turbocolor.bold.green( "Exit code: 0" ) );
      }
      logEol();
    } );

    runner.on( "start", () => {
      NodeReporter.showConcurrency( runner );
      NodeReporter.showDebuggers( runner );

      this.spinner = ora( `Waiting for "runStart"...` ).start();
    } );

    runner.on( "sigint", _try => {
      this.interrupted = true;

      if ( this.spinner ) {
        this.spinner.stop();
      }
      log( `\nStopping tests... (${_try} try)\n` );
    } );

    runner.on( "why-is-running", async why => {
      log( `\n${turbocolor.bold.red( "Resources still running in child process..." )}\n` );

      for ( const { type, stack } of why ) {
        log( `\n${type}\n` );
        await this.logDefaultByStack( stack );
      }
      logEol();
    } );
  }

  static showOptions( options ) {
    logEol();
    log( turbocolor.bold.green( "Patterns: " ) + options.files.join( " " ) + "\n" );
    if ( options.ignore.length > 0 ) {
      log( turbocolor.bold.green( "Ignore patterns: " ) + options.ignore.join( " " ) + "\n" );
    }
    logEol();
  }

  static showFilesCount( count ) {
    log( turbocolor.bold.green( "Files count: " ) + count + "\n" );
    logEol();
  }

  static showSeed( runner ) {
    if ( runner.options.random ) {
      log( turbocolor.bold.yellow( "Random seed: " ) + runner.options.random + "\n" );
      logEol();
    }
  }

  static showConcurrency( runner ) {
    log( turbocolor.bold.green( "Child processes: " ) + runner.forks.length + "\n" );
    logEol();
  }

  static showDebuggers( { debuggersPromises, division } ) {
    if ( !debuggersPromises.length ) {
      return;
    }
    Promise.all( debuggersPromises ).then( debuggers => {
      log( turbocolor.bold.yellow( "Debugging" ) );
      logEol();
      log( turbocolor.bold.yellow( "Got to chrome://inspect or check https://nodejs.org/en/docs/inspector" ) );
      logEol();
      for ( let i = 0; i < debuggers.length; i++ ) {
        log( turbocolor.bold( debuggers[ i ] ), 4 );
        logEol();
        for ( const file of division[ i ] ) {
          log( prettify( file ), 6 );
          logEol();
        }
      }
      logEol();
    } );
  }

  static fatalError( error ) {
    process.exitCode = 1;
    logEol();
    log( turbocolor.bold.red( error ) );
    logEol();
  }

  async logOtherErrors() {
    const otherErrors = this.otherErrors;
    this.otherErrors = [];

    if ( otherErrors.length > 0 ) {
      this.afterRun();
      for ( let i = 0; i < otherErrors.length; i++ ) {
        await this.logError( otherErrors[ i ] ); // eslint-disable-line no-await-in-loop
      }
    }
  }

  afterRun() {
    if ( !this.didAfterRun ) {
      this.didAfterRun = true;
      process.stdout.write( "\n  More errors:\n" );
    }
  }

  static init( runner ) {
    return new NodeReporter( runner );
  }

  runStart() {
    this.spinner.text = "Running tests...";
  }

  runEnd( t ) {
    setTimeout( async() => {

      this.spinner.stop();

      const tests = await Promise.all( this.tests );
      this.tests = null; // Prevent memory leaks

      const sortedTests = tests.sort( ( a, b ) => {
        if ( !a.source || !a.source.file ) {
          return 1;
        }
        if ( !b.source || !b.source.file ) {
          return -1;
        }
        const fileCompare = a.source.file.localeCompare( b.source.file );
        if ( fileCompare === 0 ) {
          const lineCompare = a.source.line - b.source.line;
          if ( lineCompare === 0 ) {
            return a.source.column - b.source.column;
          }
          return lineCompare;
        }
        return fileCompare;
      } );

      for ( let i = 0; i < sortedTests.length; i++ ) {
        await this.logTestEnd( sortedTests[ i ] ); // eslint-disable-line no-await-in-loop
      }

      const hasPending = t.pendingTests && t.pendingTests.size > 0;

      // If we had pending tests, don't trust the stats
      if ( t.runStartNotEmitted || hasPending ) {

        if ( t.runStartNotEmitted ) {
          log( `\n${turbocolor.bold.red( `${t.runStartNotEmitted} forks did not emit "runStart" event.` )}\n` );
        }

        if ( hasPending ) {
          log( `\n${turbocolor.bold.red( `${t.pendingTests.size} Pending tests:` )}\n` );

          for ( const stack of t.pendingTests ) {
            await this.logDefaultByStack( stack );
          }
        }

      } else {

        const { passed, skipped, todo, failed, total } = t.testCounts;

        let lines;

        if ( total === 0 ) {
          lines = [
            turbocolor.red( "\n  The total number of tests was 0." )
          ];
        } else {
          lines = [
            passed > 0 ? "\n  " + turbocolor.green( passed + " passed" ) : "",
            skipped > 0 ? "\n  " + turbocolor.yellow( skipped + " skipped" ) : "",
            todo > 0 ? "\n  " + turbocolor.blue( todo + " todo" ) : "",
            failed > 0 ? "\n  " + turbocolor.red( failed + " failed" ) : "",
            "\n\n  " + turbocolor.gray( t.runtime + " ms" ),
            t.onlyCount ? `\n\n  The '.only' modifier was used ${t.onlyCount} time${t.onlyCount === 1 ? "" : "s"}.` : ""
          ].filter( Boolean );
        }

        if ( this.runner.options.only ) {
          lines.push( "\n  --only option was used." );
        }

        if ( this.runner.options.bail ) {
          lines.push( "\n  --bail option was used." );
        }

        if ( t.snapshotStats ) {
          this.showSnapshotStats( lines, t.snapshotStats );
        }

        process.stdout.write( lines.join( "" ) + "\n" );

      }

      log( `\n${turbocolor.gray( `[${new Date().toLocaleTimeString()}]` )}\n\n` );

      await this.logOtherErrors();

      this.ended = true;

      const debuggersWaitingPromises = this.runner.debuggersWaitingPromises;

      if ( debuggersWaitingPromises.length ) {
        Promise.race( debuggersWaitingPromises ).then( () => {
          log(
            turbocolor.bold.yellow( "At least 1 of " + debuggersWaitingPromises.length + " processes are waiting for the debugger to disconnect...\n" )
          );
          logEol();
        } );
      }

    } );
  }

  showSnapshotStats( lines, { added, updated, removed, obsolete } ) {
    if ( added || updated || removed || obsolete ) {
      lines.push( turbocolor.blue( "\n\n  Snapshots" ) );
    }
    if ( added ) {
      lines.push( turbocolor.green( `\n    Added: ${added}` ) );
    }
    if ( updated ) {
      lines.push( turbocolor.green( `\n    Updated: ${updated}` ) );
    }
    if ( removed ) {
      lines.push( turbocolor.red( `\n    Removed: ${removed}` ) );
    }
    if ( obsolete ) {
      lines.push( turbocolor.red( `\n    Obsolete: ${obsolete}` ) );
    }
  }

  updateSpinner() {
    for ( const text of this.pendingTests ) {
      this.spinner.text = text;
      return;
    }
    this.spinner.text = "Waiting...";
  }

  testStart( t ) {
    this.pendingTests.add( t.fullname.join( " > " ) );
    if ( this.pendingTests.size === 1 ) {
      this.updateSpinner();
    }
  }

  testEnd( t ) {
    this.pendingTests.delete( t.fullname.join( " > " ) );
    this.updateSpinner();

    this.tests.push( ( async() => {
      const { source } = await this.runner.beautifyStack( t.defaultStack );
      t.source = source;
      return t;
    } )() );
  }

  async enhanceError( original ) {

    const err = {
      actual: null,
      expected: null,
      diff: original.diff,
      stack: null,
      source: null,
      message: original.message
    };

    // Prevent memory leaks
    original.actual = null;
    original.expected = null;

    if ( original.stack ) {
      const { stack, source } = await this.runner.beautifyStack( original.stack );
      err.stack = stack;
      err.source = source;
    }

    return err;
  }

  showSource( source ) {

    const { file, code, line, column } = source;

    if ( !file || !code ) {
      if ( file ) {
        return `${turbocolor.gray( `${prettify( file )}:${line}:${column}` )}\n`;
      }
      return "";
    }

    const codeFrameOptions = this.runner.options.codeFrameOptions;
    const frame = this.runner.options.codeFrame ?
      codeFrameColumns( code, { start: { line } }, codeFrameOptions ) + "\n\n" :
      "";

    return `${turbocolor.gray( `${prettify( file )}:${line}:${column}` )}\n\n${frame}`;
  }

  async logDefaultByStack( defaultStack ) {
    const { source } = await this.runner.beautifyStack( defaultStack );
    this.logDefault( source );
  }

  logDefault( source ) {
    let log = "\n";

    if ( source ) {
      log += this.showSource( source );
    }

    printLog( log, 4 );
  }

  makeLegend() {
    const { diffGutters } = this.runner.options.concordanceOptions.theme;
    return `${diffGutters.actual}Actual ${diffGutters.expected}Expected`;
  }

  async logError( e ) {

    const diff = this.runner.options.diff;
    const stack = this.runner.options.stack;

    const error = await this.enhanceError( e );
    let log = "\n";

    if ( error.message ) {
      log += turbocolor.bold( error.message ) + "\n";
    }

    if ( error.source ) {
      log += this.showSource( error.source );
    }

    if ( diff && error.diff ) {
      log += `${this.makeLegend()}\n\n${indentString( error.diff )}\n\n`;
    }

    if ( stack && error.stack ) {
      log += turbocolor.gray( error.stack ) + "\n\n";
    }

    printLog( log, 4 );
  }

  async logTestEnd( { fullname, status, skipReason, errors, runtime, slow, logs, source, memoryUsage } ) {

    if ( !this.runner.options.verbose || !logs.length ) {
      if ( status === "passed" && !slow ) {
        return;
      }
      if ( skipReason === skipReasons.bailed ) {
        return;
      }
      if ( skipReason === skipReasons.interrupted ) {
        return;
      }
    }

    const statusText = status === "failed" ? turbocolor.red( status ) : status === "passed" ? turbocolor.green( status ) : turbocolor.yellow( status );

    printLog( `\n${turbocolor.bold( fullname.join( " > " ) )}\n${statusText} | ${runtime} ms ${slow ? turbocolor.yellow( "Slow!" ) : ""}\n` );

    if ( skipReason ) {
      printLog( `\nSkip reason: ${skipReason}`, 4 );
    }

    if ( errors.length ) {
      for ( let i = 0; i < errors.length; i++ ) {
        await this.logError( errors[ i ] ); // eslint-disable-line no-await-in-loop
      }
    } else {
      this.logDefault( source );
    }

    if ( logs.length ) {

      printLog( "Logs:\n\n", 4 );

      logs.forEach( log => {
        const logLines = indentString( turbocolor.gray( log ), 6 );
        const logLinesWithFigure = logLines.replace(
          /^ {6}/,
          `    ${logSymbols.info} `
        );

        printLog( logLinesWithFigure, 0 );
        logEol();
      } );
    }

    if ( memoryUsage ) {
      printLog( turbocolor.yellow( `Memory usage: ${memoryUsage}\n` ), 4 );
      logEol();
    }

  }

}
