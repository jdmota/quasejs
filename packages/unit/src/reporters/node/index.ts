import turbocolor from "turbocolor";
import logSymbols from "log-symbols";
import ora from "ora";
import { codeFrameColumns } from "@babel/code-frame";
import { prettify } from "@quase/path-url";
import skipReasons from "../../core/skip-reasons";
import { TestEnd, RunEnd, SnapshotStats, TestStart, EventAskSourceContent } from "../../types";
import { log, log as printLog, logEol, indentString } from "./log";
import { NodeRunner } from "../../node-runner";

export default class NodeReporter {

  private runner: NodeRunner;
  private spinner: ora.Ora | null;
  private pendingTests: Set<string>;
  private tests: Promise<TestEnd>[];
  private otherErrors: any[];
  private ended: boolean;
  private didAfterRun: boolean;
  private interrupted: boolean;

  constructor( runner: NodeRunner ) {
    this.runner = runner;
    this.spinner = null;
    this.pendingTests = new Set();
    this.tests = [];
    this.otherErrors = [];
    this.ended = false;
    this.didAfterRun = false;
    this.interrupted = false;
    runner.on( "runStart", this.runStart.bind( this ) );
    runner.on( "testStart", this.testStart.bind( this ) );
    runner.on( "testEnd", this.testEnd.bind( this ) );
    runner.on( "runEnd", this.runEnd.bind( this ) );
    runner.on( "otherError", err => {
      this.otherErrors.push( err );
      if ( this.ended ) {
        this.logOtherErrors();
      }
    } );

    runner.on( "global-timeout", () => {
      this.getSpinner().stop();

      log( `\n${turbocolor.bold.red( "Global timeout!" )}\n\n` );
    } );

    runner.on( "exit", async() => {
      this.getSpinner().stop();

      NodeReporter.showSeed( runner );

      await this.logOtherErrors();

      if ( this.interrupted ) {
        log( `\n${turbocolor.bold.red( "Interrupted." )}\n\n` );
      }

      if ( process.exitCode ) {
        log( turbocolor.bold.red( "Exit code: " + process.exitCode ) );
      } else {
        log( turbocolor.bold.green( "Exit code: 0" ) );
      }
      logEol();
    } );

    runner.on( "start", () => {
      NodeReporter.showConcurrency( runner );
      NodeReporter.showDebuggers( runner );

      this.spinner = ora( `Waiting for "runStart"...` ).start();
    } );

    runner.on( "sigint", _try => {
      this.interrupted = true;

      if ( this.spinner ) {
        this.spinner.stop();
      }
      log( `\nStopping tests... (${_try} try)\n` );
    } );

    runner.on( "why-is-running", async why => {
      log( `\n${turbocolor.bold.red( "Resources still running in child process..." )}\n` );

      for ( const { type, stack } of why ) {
        log( `\n${type}\n` );
        await this.logDefault( stack );
      }
      logEol();
    } );
  }

  static showOptions( options: any ) {
    logEol();
    log( turbocolor.bold.green( "Patterns: " ) + options.files.join( " " ) + "\n" );
    if ( options.ignore.length > 0 ) {
      log( turbocolor.bold.green( "Ignore patterns: " ) + options.ignore.join( " " ) + "\n" );
    }
    logEol();
  }

  static showFilesCount( count: number, time: number ) {
    log( `${turbocolor.bold.green( `Found ${count} files` )} in ${time} ms\n` );
    logEol();
  }

  static showSeed( runner: NodeRunner ) {
    if ( runner.options.random ) {
      log( turbocolor.bold.yellow( "Random seed: " ) + runner.options.random + "\n" );
      logEol();
    }
  }

  static showConcurrency( runner: NodeRunner ) {
    log( turbocolor.bold.green( "Child processes: " ) + runner.forks.length + "\n" );
    logEol();
  }

  static showDebuggers( runner: NodeRunner ) {
    if ( !runner.debuggersPromises.length ) {
      return;
    }
    Promise.all( runner.debuggersPromises ).then( debuggers => {
      log( turbocolor.bold.yellow( "Debugging" ) );
      logEol();
      log( turbocolor.bold.yellow( "Got to chrome://inspect or check https://nodejs.org/en/docs/inspector" ) );
      logEol();
      for ( let i = 0; i < debuggers.length; i++ ) {
        log( turbocolor.bold( debuggers[ i ] ), 4 );
        logEol();
        for ( const file of runner.division[ i ] ) {
          log( prettify( file ), 6 );
          logEol();
        }
      }
      logEol();
    } );
  }

  static fatalError( error: string ) {
    process.exitCode = 1;
    logEol();
    log( turbocolor.bold.red( error ) );
    logEol();
  }

  getSpinner() {
    const { spinner } = this;
    if ( spinner ) {
      return spinner;
    }
    throw new Error( `No spinner` );
  }

  async logOtherErrors() {
    const otherErrors = this.otherErrors;
    this.otherErrors = [];

    if ( otherErrors.length > 0 ) {
      this.afterRun();
      for ( let i = 0; i < otherErrors.length; i++ ) {
        await this.logError( otherErrors[ i ] ); // eslint-disable-line no-await-in-loop
      }
    }
  }

  afterRun() {
    if ( !this.didAfterRun ) {
      this.didAfterRun = true;
      process.stdout.write( "\n  More errors:\n" );
    }
  }

  static init( runner: NodeRunner ) {
    return new NodeReporter( runner );
  }

  runStart() {
    this.getSpinner().text = "Running tests...";
  }

  runEnd( t: RunEnd ) {
    setTimeout( async() => {

      this.getSpinner().stop();

      const tests = await Promise.all( this.tests );
      this.tests = []; // Prevent memory leaks

      const sortedTests = tests.sort( ( a, b ) => {
        if ( !a.source || !a.source.file ) {
          return 1;
        }
        if ( !b.source || !b.source.file ) {
          return -1;
        }
        const fileCompare = a.source.file.localeCompare( b.source.file );
        if ( fileCompare === 0 ) {
          if ( !a.source.line || !b.source.line ) {
            return 0;
          }
          const lineCompare = a.source.line - b.source.line;
          if ( lineCompare === 0 ) {
            if ( !a.source.column || !b.source.column ) {
              return 0;
            }
            return a.source.column - b.source.column;
          }
          return lineCompare;
        }
        return fileCompare;
      } );

      for ( let i = 0; i < sortedTests.length; i++ ) {
        await this.logTestEnd( sortedTests[ i ] ); // eslint-disable-line no-await-in-loop
      }

      const notStartedForks = t.notStartedForks;
      const hasPending = t.pendingTests && t.pendingTests.size > 0;

      // If we had pending tests, don't trust the stats
      if ( notStartedForks.length > 0 || hasPending ) {

        if ( notStartedForks.length > 0 ) {
          log( `\n${turbocolor.bold.red(
            `${notStartedForks.length} child ${notStartedForks.length === 1 ? "process" : "processes"}` +
            ` did not emit "runStart" event.`
          )}\n` );
          log( `\nPossible causes:\n` );

          for ( const { didCrash, notImportedFiles } of notStartedForks ) {
            if ( didCrash || notImportedFiles.size === 0 ) {
              log( `  - Some error during setup. See below.\n` );
            } else {
              for ( const file of notImportedFiles ) {
                log( `  - Infinite loop in ${prettify( file )}\n` );
                break; // Only show the first one, since files are imported by order
              }
            }
          }
        }

        if ( hasPending ) {
          log( `\n${turbocolor.bold.red( `${t.pendingTests.size} Pending tests:` )}\n` );

          for ( const stack of t.pendingTests ) {
            await this.logDefault( stack );
          }
        }

      } else {

        const { passed, skipped, todo, failed, total } = t.testCounts;

        let lines;

        if ( total === 0 ) {
          lines = [
            turbocolor.red( "\n  The total number of tests was 0." )
          ];
        } else {
          lines = [
            passed > 0 ? "\n  " + turbocolor.green( passed + " passed" ) : "",
            skipped > 0 ? "\n  " + turbocolor.yellow( skipped + " skipped" ) : "",
            todo > 0 ? "\n  " + turbocolor.blue( todo + " todo" ) : "",
            failed > 0 ? "\n  " + turbocolor.red( failed + " failed" ) : "",
            "\n\n  " + turbocolor.gray( t.runtime + " ms" ),
            t.onlyCount ? `\n\n  The '.only' modifier was used ${t.onlyCount} time${t.onlyCount === 1 ? "" : "s"}.` : ""
          ].filter( Boolean );
        }

        if ( this.runner.options.only ) {
          lines.push( "\n  --only option was used." );
        }

        if ( this.runner.options.bail ) {
          lines.push( "\n  --bail option was used." );
        }

        if ( t.snapshotStats ) {
          this.showSnapshotStats( lines, t.snapshotStats );
        }

        process.stdout.write( lines.join( "" ) + "\n" );

      }

      log( `\n${turbocolor.gray( `[${new Date().toLocaleTimeString()}]` )}\n\n` );

      await this.logOtherErrors();

      this.ended = true;

      const debuggersWaitingPromises = this.runner.debuggersWaitingPromises;

      if ( debuggersWaitingPromises.length ) {
        Promise.race( debuggersWaitingPromises ).then( () => {
          log(
            turbocolor.bold.yellow( "At least 1 of " + debuggersWaitingPromises.length + " processes are waiting for the debugger to disconnect...\n" )
          );
          logEol();
        } );
      }

    } );
  }

  showSnapshotStats( lines: string[], { added, updated, removed, obsolete }: SnapshotStats ) {
    if ( added || updated || removed || obsolete ) {
      lines.push( turbocolor.blue( "\n\n  Snapshots" ) );
    }
    if ( added ) {
      lines.push( turbocolor.green( `\n    Added: ${added}` ) );
    }
    if ( updated ) {
      lines.push( turbocolor.green( `\n    Updated: ${updated}` ) );
    }
    if ( removed ) {
      lines.push( turbocolor.red( `\n    Removed: ${removed}` ) );
    }
    if ( obsolete ) {
      lines.push( turbocolor.red( `\n    Obsolete: ${obsolete}` ) );
    }
  }

  updateSpinner() {
    const spinner = this.getSpinner();
    for ( const text of this.pendingTests ) {
      spinner.text = text;
      return;
    }
    spinner.text = "Waiting...";
  }

  testStart( t: TestStart ) {
    this.pendingTests.add( t.fullname.join( " > " ) );
    if ( this.pendingTests.size === 1 ) {
      this.updateSpinner();
    }
  }

  testEnd( t: TestEnd ) {
    this.pendingTests.delete( t.fullname.join( " > " ) );
    this.updateSpinner();

    this.tests.push( ( async() => {
      const { source } = await this.runner.beautifyStack( t.defaultStack );
      t.source = source;
      return t;
    } )() );
  }

  async enhanceError( original: any ): Promise<any> {

    const err: any = {
      actual: null,
      expected: null,
      diff: original.diff,
      stack: null,
      source: null,
      message: original.message
    };

    // Prevent memory leaks
    original.actual = null;
    original.expected = null;

    if ( original.stack ) {
      const { stack, source } = await this.runner.beautifyStack( original.stack );
      err.stack = stack;
      err.source = source;
    }

    return err;
  }

  showSource( source: EventAskSourceContent ) {

    const { file, code, line, column } = source;

    if ( !file || !code ) {
      if ( file ) {
        return `${turbocolor.gray( `${prettify( file )}:${line}:${column}` )}\n`;
      }
      return "";
    }

    const codeFrameOptions = this.runner.options.codeFrameOptions;
    const frame = this.runner.options.codeFrame ?
      codeFrameColumns( code, { start: { line } }, codeFrameOptions ) + "\n\n" :
      "";

    return `${turbocolor.gray( `${prettify( file )}:${line}:${column}` )}\n\n${frame}`;
  }

  async logDefault( defaultStack: string, source?: EventAskSourceContent | null ) {
    if ( !source ) {
      source = ( await this.runner.beautifyStack( defaultStack ) ).source;
    }

    let log = "\n";

    if ( source ) {
      log += this.showSource( source );
    }

    printLog( log, 4 );
  }

  makeLegend() {
    const { diffGutters } = this.runner.options.concordanceOptions.theme;
    return `${diffGutters.actual}Actual ${diffGutters.expected}Expected`;
  }

  async logError( e: any ) {

    const diff = this.runner.options.diff;
    const stack = this.runner.options.stack;

    const error = await this.enhanceError( e );
    let log = "\n";

    if ( error.message ) {
      log += turbocolor.bold( error.message ) + "\n";
    }

    if ( error.source ) {
      log += this.showSource( error.source );
    }

    if ( diff && error.diff ) {
      log += `${this.makeLegend()}\n\n${indentString( error.diff )}\n\n`;
    }

    if ( stack && error.stack ) {
      log += turbocolor.gray( error.stack ) + "\n\n";
    }

    printLog( log, 4 );
  }

  async logTestEnd( { defaultStack, fullname, status, skipReason, errors, runtime, slow, logs, memoryUsage, source }: TestEnd ) {

    if ( !this.runner.options.verbose || !logs.length ) {
      if ( status === "passed" && !slow ) {
        return;
      }
      if ( skipReason === skipReasons.bailed ) {
        return;
      }
      if ( skipReason === skipReasons.interrupted ) {
        return;
      }
    }

    const statusText = status === "failed" ? turbocolor.red( status ) : status === "passed" ? turbocolor.green( status ) : turbocolor.yellow( status );

    printLog( `\n${turbocolor.bold( fullname.join( " > " ) )}\n${statusText} | ${runtime} ms ${slow ? turbocolor.yellow( "Slow!" ) : ""}\n` );

    if ( skipReason ) {
      printLog( `\nSkip reason: ${skipReason}`, 4 );
    }

    if ( errors.length ) {
      for ( let i = 0; i < errors.length; i++ ) {
        await this.logError( errors[ i ] ); // eslint-disable-line no-await-in-loop
      }
    } else {
      await this.logDefault( defaultStack, source );
    }

    if ( logs.length ) {

      printLog( "Logs:\n\n", 4 );

      logs.forEach( log => {
        const logLines = indentString( turbocolor.gray( log ), 6 );
        const logLinesWithFigure = logLines.replace(
          /^ {6}/,
          `    ${logSymbols.info} `
        );

        printLog( logLinesWithFigure, 0 );
        logEol();
      } );
    }

    if ( memoryUsage ) {
      printLog( turbocolor.yellow( `Memory usage: ${memoryUsage}\n` ), 4 );
      logEol();
    }

  }

}
