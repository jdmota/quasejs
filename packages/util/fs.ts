import fs from "fs/promises";

// https://nodejs.org/api/errors.html#nodejs-error-codes

// "No such file or directory"
export function isNoFileError(error: any) {
  return error != null && error.code === "ENOENT";
}

export async function readFileString(filename: string): Promise<string> {
  try {
    return await fs.readFile(filename, "utf8");
  } catch (error) {
    if (isNoFileError(error)) {
      return "";
    }
    throw error;
  }
}
