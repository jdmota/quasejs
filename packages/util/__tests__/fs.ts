import { expect, it } from "@jest/globals";
import { resolve } from "node:path";
import { findUp, isFile, readFileString } from "../fs";

const fixtures = resolve(import.meta.dirname, "__fixtures__");

it("readFileString", async () => {
  await expect(readFileString(resolve(fixtures, "a.txt"))).resolves.toBe("a");
  await expect(readFileString(resolve(fixtures, "empty.txt"))).resolves.toBe(
    ""
  );
  await expect(readFileString(resolve(fixtures, "non-existant"))).resolves.toBe(
    ""
  );
});

it("isFile", async () => {
  await expect(isFile(import.meta.filename, true)).resolves.toBe(true);
  await expect(isFile(resolve(fixtures, "a.txt"), true)).resolves.toBe(true);
  await expect(isFile(resolve(fixtures, "non-existant"), true)).resolves.toBe(
    false
  );
});

it("findUp", async () => {
  await expect(
    findUp(["a.txt", "b.txt", "c.txt"], {
      cwd: fixtures,
    })
  ).resolves.toMatchSnapshot();

  await expect(
    findUp(["a.txt", "b.txt", "c.txt"], {
      cwd: fixtures,
      multipleAtLevel: true,
    })
  ).resolves.toMatchSnapshot();

  await expect(
    findUp(["inner-a.txt", "a.txt", "b.txt", "c.txt"], {
      cwd: resolve(fixtures, "folder"),
      stopAt: fixtures,
      multipleAtLevel: true,
      multipleUp: true,
    })
  ).resolves.toMatchSnapshot();
});
