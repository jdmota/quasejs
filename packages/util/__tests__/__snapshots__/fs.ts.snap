// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`findUp 1`] = `
[
  "packages/util/__tests__/__fixtures__/a.txt",
]
`;

exports[`findUp 2`] = `
[
  "packages/util/__tests__/__fixtures__/a.txt",
  "packages/util/__tests__/__fixtures__/b.txt",
]
`;

exports[`findUp 3`] = `
[
  "packages/util/__tests__/__fixtures__/folder/inner-a.txt",
  "packages/util/__tests__/__fixtures__/a.txt",
  "packages/util/__tests__/__fixtures__/b.txt",
]
`;
